lib/resources/etc_hosts_allow_deny.rb

```diff
@@ -0,0 +1,162 @@
+# encoding: utf-8
+# author: Matthew Dromazos
+
+require 'utils/parser'
+
+module Inspec::Resources
+  class EtcHostsAllow < Inspec.resource(1)
+    name 'etc_hosts_allow'
+    desc 'Use the etc_hosts_allow InSpec audit resource to test the connections
+          the client will allow. Controlled by the /etc/hosts.allow file.'
+    example "
+      describe etc_hosts_allow.where { daemon == 'ALL' } do
+        its('client_list') { should include ['127.0.0.1', '[::1]'] }
+        its('options') { should eq [[]] }
+      end
+    "
+
+    attr_reader :params
+
+    include CommentParser
+
+    def initialize(hosts_allow_path = nil)
+      return skip_resource 'The `etc_hosts_allow` resource is not supported on your OS.' unless inspec.os.linux?
+      @conf_path      = hosts_allow_path || '/etc/hosts.allow'
+      @files_contents = {}
+      @content        = nil
+      @params         = nil
+      read_content
+    end
+
+    filter = FilterTable.create
+    filter.add_accessor(:where)
+          .add_accessor(:entries)
+          .add(:daemon,      field: 'daemon')
+          .add(:client_list, field: 'client_list')
+          .add(:options,     field: 'options')
+
+    filter.connect(self, :params)
+
+    private
+
+    def read_content
+      @content = ''
+      @params  = {}
+      @content = split_daemons(read_file(@conf_path))
+      @params  = parse_conf(@content)
+    end
+
+    def split_daemons(content)
+      split_daemons_list = []
+      content.each do |line|
+        data, = parse_comment_line(line, comment_char: '#', standalone_comments: false)
+        next unless data != ''
+        data.split(':')[0].split(',').each do |daemon|
+          split_daemons_list.push("#{daemon} : " + line.split(':', 2)[1])
+        end
+      end
+      split_daemons_list
+    end
+
+    def parse_conf(content)
+      content.map do |line|
+        parse_line(line) unless line == ''
+      end.compact
+    end
+
+    def parse_line(line)
```

Comment:
While this method, and the `parse_attributes` method, are creative ways of solving the parsing problem, I think they may be too complicated and there's definitely a high maintenance burden if we have to debug it in the future.

I think this can be a lot simpler. I whipped this method up for some local testing and I think something similar to this would do the trick:

```ruby
def parse(line)
    daemons, clients_and_options = line.split(/\s+:\s+/, 2)
    daemon_list = daemons.split(',').map(&:strip)

    clients_and_options ||= ''
    clients, options = clients_and_options.split(' : ', 2)
    client_list = clients.split(',').map(&:strip)

    options ||= ''
    options_list = options.split(': ').map(&:strip)

    puts "Daemons: #{daemon_list.inspect}"
    puts "Clients: #{client_list.inspect}"
    puts "Options: #{options_list.inspect}"
end
```

I'd also like to see some specific unit tests for the method that parses these lines given how complex they can be. The tests should focus on parsing a single line that you hardcode in your tests. The tests might look something like this:

```ruby
describe 'Inspec::Resources::EtcHostsAllow' do
  describe '#parse_line' do
    it 'parses a line with multiple clients' do
      line = 'sshd : 1.2.3.4, 4.3.2.1'
      _(resource.parse_line(line)['daemons']).must_equal ['sshd']
      _(resource.parse_line(line)['clients']).must_equal ['1.2.3.4', '4.3.2.1']
    end

    it 'parses a line with multiple daemons' do
      # etc
    end

    it 'parses a line with one option' do
    end

    it 'parses a line with multiple options' do
    end
  end
end
```

Let me know if you have questions about this or if you think my suggestion doesn't properly parse these lines.

I will note that hosts.allow / hosts.deny files can contain entries that span multiple lines and can even reference other files for the client list, for example. We don't have to solve that problem here, but it's something to be aware of and why I want the line parsing methods to be as simple as can be.

Reply:
The only issue I see with this is when you have an ipv6 address in the client list, because then you cannot split by ':'

Reply:
It still works because the tcpwrappers spec states that there must be a space before and after the `:` between the clients and any options.

```